extension/intents/tabs/tabs.js

```diff
@@ -371,3 +371,179 @@ intentRunner.registerIntent({
     await browser.tabs.move(tabsIds, { windowId: newWindow.id, index: 0 });
   },
 });
+
+async function collectTabsAround(center, tabs) {
+  if (tabs === undefined) {
+    return;
+  }
+
+  const tabsLeft = tabs.filter(tab => {
+    return tab.index < center.index;
+  });
+
+  const tabsRight = tabs.filter(tab => {
+    return tab.index > center.index;
+  });
+
+  for (let i = 0; i < tabsLeft.length; i++) {
+    await browser.tabs.move(tabsLeft[tabsLeft.length - i - 1].id, {
+      index: center.index - i - 1,
+    });
+  }
+
+  for (let i = 0; i < tabsRight.length; i++) {
+    await browser.tabs.move(tabsRight[i].id, { index: center.index + i + 1 });
+  }
+}
+
+async function collectRightTabs(tabs) {
+  if (tabs === undefined || tabs.length === 0) {
+    return;
+  }
+
+  const leftestTab = tabs[0];
+  const tabsRight = tabs.slice(1);
+
+  for (let i = 0; i < tabsRight.length; i++) {
+    await browser.tabs.move(tabsRight[i].id, {
+      index: leftestTab.index + i + 1,
+    });
+  }
+}
+
+function groupTabs(tabs, keyGetter) {
+  if (tabs === undefined) {
+    return {};
+  }
+
+  const groupedTabs = {};
+
+  for (let i = 0; i < tabs.length; i++) {
+    const key = keyGetter(tabs[i]);
+
+    if (groupedTabs[key] === undefined) {
+      groupedTabs[key] = [tabs[i]];
+    } else {
+      groupedTabs[key].push(tabs[i]);
+    }
+  }
+
+  return groupedTabs;
+}
+
+async function getMatchingTabs(options) {
+  const tabQuery = {
+    pinned: false,
+    windowType: "normal",
+  };
+
+  if (options.currentWindow === "true") {
+    tabQuery.currentWindow = true;
+  }
+
+  if (options.query === undefined && options.activeTab !== undefined) {
+    const activeTab = options.activeTab;
+    const activeHostname = new URL(activeTab.url).hostname;
+    tabQuery.url = "*://*." + activeHostname + "/*";
+  }
+
+  let matchingTabs = await browser.tabs.query(tabQuery);
+
+  if (options.query !== undefined) {
+    matchingTabs = matchingTabs.filter(tab => {
+      return new URL(tab.url).origin.includes(options.query);
```

Comment:
This should also test the title, and we'll want a case-insensitive test. Arguably we should be using Fuse for searching (or some other library per #1089), but I think for now a simple string search is OK and adding fuzzy search can be a followup ticket.